### PR TITLE
[Merged by Bors] - Prepare proposer

### LINKED ISSUE #2936: Provide `payloadAttributes` for block proposers in advance
## Description

[`engine_forkchoiceUpdatedV1`]: https://github.com/ethereum/execution-apis/blob/main/src/engine/specification.md#engine_forkchoiceupdatedv1
[`engine_getPayloadV1`]: https://github.com/ethereum/execution-apis/blob/main/src/engine/specification.md#engine_getpayloadv1

According to the [execution API spec](https://github.com/ethereum/execution-apis/blob/main/src/engine/specification.md), consensus clients must call [`engine_forkchoiceUpdatedV1`][] (fcU) with a `payloadAttributes` that is non-`null` when the node expects to produce a block in the next slot. This gives the execution engine time to prepare the optimal set of transactions, before we request the payload.

Lighthouse is not presently sending the `payloadAttributes` in advance. Instead, it always ends up hitting the [fall-back path](https://github.com/sigp/lighthouse/blob/ef7351ddfecbb35cd23dde11835d58d4d4e689a8/beacon_node/execution_layer/src/lib.rs#L276-L300) which sends the fcU message immediately before the [`engine_getPayloadV1`][] call.

## Mechanisms

In order to supply `payloadAttributes` in advance, we need the VC to call the `/eth/v1/validator/prepare_beacon_proposer`, an endpoint currently implemented in a [PR to the beacon-APIs repo](https://github.com/ethereum/beacon-APIs/pull/178).

The simplest implementation would be for the BN to immediately do an fcU call when it receives something on the `prepare_beacon_proposer` endpoint. However, I don't think that's how we should implement it. There's a whole bunch of edge-cases regarding re-orgs and timing assumptions, I think it's best if the VC just says "hey I have this validator" and then the BN takes care of issuing the right call at the right time. AFAIK, this is how all other clients are implementing it too.

To implement it properly, I think we should build a service in the BN that accepts `prepare_beacon_proposer` calls and then "registers" a `validator_index` as a local, potential block producer. Then, that service would issues the fcU with the appropriate `payloadAttributes` when it's some time (`t`) before the start of a slot. It would also be re-org aware and could make sure the execution engine is kept up-to-date relative to a changing head.

The API docs state:

> The information supplied for each validator index will persist through the epoch in which the call is submitted and for a further two epochs after that, or until the beacon node restarts.

So, this service needs to track each validator for 2-3 epochs and then forget about it. That means that the VC will need to issue this call *at least* once every 2-3 epochs. I suggest we do it more frequently, perhaps every epoch or every slot.

## Steps Required

As I see it, we need to do the following things:

1. Add the mechanisms to the VC to poll the `prepare_beacon_proposer` with all its validators (that have validator indices) every slot/epoch/whatever.
1. Add the service to the BN that can track the registered validator indices and issue the fcU calls at the appropriate times.
    - *Note: we can call fcU on the EL whenever we want, we don't have to make sure we call it when our head changes. We can call it with the same head if we like.*
1. Build tests

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

Resolves #2936

## Proposed Changes

Adds functionality for calling [`validator/prepare_beacon_proposer`](https://ethereum.github.io/beacon-APIs/?urls.primaryName=dev#/Validator/prepareBeaconProposer) in advance.

There is a `BeaconChain::prepare_beacon_proposer` method which, which called, computes the proposer for the next slot. If that proposer has been registered via the `validator/prepare_beacon_proposer` API method, then the `beacon_chain.execution_layer` will be provided the `PayloadAttributes` for us in all future forkchoiceUpdated calls. An artificial forkchoiceUpdated call will be created 4s before each slot, when the head updates and when a validator updates their information.

Additionally, I added strict ordering for calls from the `BeaconChain` to the `ExecutionLayer`. I'm not certain the `ExecutionLayer` will always maintain this ordering, but it's a good start to have consistency from the `BeaconChain`. There are some deadlock opportunities introduced, they are documented in the code.

## Additional Info

- ~~Blocked on #2837~~

### CODE REVIEW

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (realbigsean) [inline beacon_node/beacon_chain/src/beacon_chain.rs:3680]: What's the benefit of running this after `update_execution_engine_forkchoice_blocking`? This seems intentional, but I am only really seeing the potential for duplicate `notify_forkchoice_updated` requests, which like you point out doesn't seem *bad* but I'm not seeing why it might be useful

File: beacon_node/beacon_chain/src/proposer_prep_service.rs
Comment (realbigsean) [inline beacon_node/beacon_chain/src/proposer_prep_service.rs:31]: It seems like this service covers the scenario where we have a skip slot right before we propose, and want to make sure we still notify the execution engine of our `PayloadAttributes`. But shouldn't that already be covered by the `state_advance_timer` which triggers at the same time and calls `fork_choice` which calls `prepare_beacon_proposer_async`?

File: beacon_node/http_api/src/lib.rs
Comment (realbigsean) [inline beacon_node/http_api/src/lib.rs:2232]: I sort of think this endpoint should only handle updating the proposer preparation cache, and we should make sure fork choice is the only place we call `prepare_beacon_proposer_async` from.  

I could see the benefit of this potentially allowing last second suggested fee recipients changes.  But I'm not sure that's something we should try to code for.  To me it seems like the later this information is provided the less profitable the block could be and the more likely it is to just be missed completely, so there's already strong incentives to provide the fee recipient early and not to change it frequently.

If we are worried about not being able to get the fee recipient on startup for validators proposing immediately, we could also persist this cache or allow the beacon chain to read from the suggested fee recipients file.

**Review by realbigsean — COMMENTED** [review]
I had a couple comments around simplifying this, but let me know if there are edge cases I am missing that are informing the design here.

Generally I was thinking the easiest/safest way to maintain BN <> EL consistency is to minimize when we call `notify_forkchoice_updated` and to only call it from `fork_choice`.  We'd be relying on fork choice's ordering which seems like the correct thing to do to me but please correct me here otherwise. 

The one place outside of fork choice where calling `notify_forkchoice_updated` seems unavoidable is when we fallback from `get_payloadV1` due to a payload id missing from our cache ([here](https://github.com/sigp/lighthouse/blob/unstable/beacon_node/execution_layer/src/lib.rs#L435)). But calling a synthetic fork choice update here doesn't really seem like a good thing anyways... could we just run a call to `chain.fork_choice()` in this scenario instead?

If we're able to only call `notify_forkchoice_updated` from fork choice, I think we could then remove all the `notify_forkchoice_updated` specific locking that's added here as well.

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (paulhauner) [reply beacon_node/beacon_chain/src/beacon_chain.rs:3680]: So `prepare_beacon_proposer_blocking` will only send a fcU (forkchciceUpdated) message to the EL if something has changed or if the timing is right.

As it's currently written, we need to call `prepare_beacon_proposer_blocking` and `notify_forkchoice_updated` to ensure that at least one fcU call goes through to the EE.

We could probably avoid this by breaking out the fcU call from `prepare_beacon_proposer_blocking` it get it to return a "should call fcU bool". I'm not certain it's worth it, but I'll give it some thought.

File: beacon_node/beacon_chain/src/proposer_prep_service.rs
Comment (paulhauner) [reply beacon_node/beacon_chain/src/proposer_prep_service.rs:31]: Good thinking, but unfortunately the `state_advance_timer` only moves forward a single slot:

https://github.com/sigp/lighthouse/blob/c3a793fd73a3b11b130b82032904d39c952869e4/beacon_node/beacon_chain/src/state_advance_timer.rs#L230-L237

So after one skip slot and following slots won't be advanced.

**Comment by paulhauner** [comment]
> Generally I was thinking the easiest/safest way to maintain BN <> EL consistency is to minimize when we call notify_forkchoice_updated and to only call it from fork_choice. 

Even if we do call it inside fork choice, we still need to deal with ordering. Even if the call to fcU is inside the `fork_choice_internal` function, it still doesn't mean the calls are ordered. Something can call `fork_choice_internal` multiple times, unless we've prevented concurrency internally via some lock. We *could* use the main `chain.fork_choice` lock, however I'm hesitant to hold that lock whilst we call out to remote servers. That is likely to cause long-lived contention.

I'm open to just returning a bool from `prepare_beacon_proposer_blocking` and callers can use that value to call fork choice, if they want. However, I think there's still downsides to that. For example, during `fork_choice_internal`, it seems valuable to me to call fcU as quickly as possible, *then* compute the proposers, *then* maybe call it again. Some ELs might avoid syncing non-head-chains until we provide them a fcU message saying it's the head. In those cases, we want the EL to be aware of the head ASAP so they can start syncing it. In the case where the shuffling isn't in the cache, it seems ideal to send 2x fcU messages; one that alerts about the head change and a later one that informs about the proposer.

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (realbigsean) [reply beacon_node/beacon_chain/src/beacon_chain.rs:3680]: Calling fcU before calculating the proposer in order to notify the EE as soon as possible seems reason enough to me to keep it as is 👌

File: beacon_node/beacon_chain/src/proposer_prep_service.rs
Comment (realbigsean) [reply beacon_node/beacon_chain/src/proposer_prep_service.rs:31]: Oh gotcha, didn't realize this

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (realbigsean) [inline beacon_node/beacon_chain/src/beacon_chain.rs:3982]: Does this need to be de-aliased? Asking because it ended up not being necessary in the recent optimistic sync PR but am not sure why

File: beacon_node/beacon_chain/src/proposer_prep_service.rs
Comment (realbigsean) [inline beacon_node/beacon_chain/src/proposer_prep_service.rs:39]: If `prepare_beacon_proposer_async` takes 4s to complete, this will wait ~20s to run again, but maybe that's ok?

File: beacon_node/execution_layer/src/lib.rs
Comment (realbigsean) [inline beacon_node/execution_layer/src/lib.rs]:
```suggestion
    /// The block will be built atop `head_block_root` and the EL will need to prepare an
```

**Comment by realbigsean** [comment]
> Even if we do call it inside fork choice, we still need to deal with ordering.  

Yes I was misunderstanding this. I was looking for a way around the mutex lock on fcU because it'll be called from a lot of different places, but I guess it is the only way to guarantee order.

**Comment by realbigsean** [comment]
I think [this call](https://github.com/sigp/lighthouse/blob/unstable/beacon_node/execution_layer/src/lib.rs#L435) also needs to use `update_execution_engine_forkchoice_async` to guarantee ordering

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (realbigsean) [inline beacon_node/beacon_chain/src/beacon_chain.rs:3775]: Will this get hit during sync?

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (paulhauner) [reply beacon_node/beacon_chain/src/beacon_chain.rs:3982]: Ah yes, good question!

We didn't need to do it in the last PR since we were getting the root from fork choice itself. Fork choice does the de-aliasing for us, so no need there. Here we're getting the root from the `BeaconState` which doesn't do the de-aliasing. I didn't notice this at first and found it in a test!

File: beacon_node/beacon_chain/src/proposer_prep_service.rs
Comment (paulhauner) [reply beacon_node/beacon_chain/src/proposer_prep_service.rs:39]: Good point! In 5ddeae81c I've made it so we spawn the task rather than waiting for it to compete. That should ensure a regular interval in all reasonable scenarios.

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (paulhauner) [reply beacon_node/beacon_chain/src/beacon_chain.rs:3775]: Great point! Resolved in 68e464887.

**Comment by paulhauner** [comment]
> I think [this call](https://github.com/sigp/lighthouse/blob/unstable/beacon_node/execution_layer/src/lib.rs#L435) also needs to use `update_execution_engine_forkchoice_async` to guarantee ordering

I agree, I don't think we can guarantee ordering without adding something there. However, it's complicated because we're producing a payload and using the fcU call only to obtain a payload ID when we don't already have one. Like you mentioned, it's possible that the head changes midway through producing a block and that fcU call is out-of-order. In such a case, you'd expect us to issue the latest head during `get_payload`, however that wouldn't help us get a payload ID for the block we're trying to produce.

So, it seems the choice during `get_payload` is to either:

1. Always send an fork choice update that's potentially out-of-order in order to obtain a payload ID.
1. Only send a fork choice update to get a payload ID *if* it agrees with the current head.

Presently we have (1) and I think that's ideal since we don't risk failing on block production. The downside of using (1) over (2) is that our head might bounce around a little from the EL perspective, which is safe but not ideal.

So, I'm tempted to leave it as-is but I'm definitely keen to hear your thoughts.

**Comment by paulhauner** [comment]
Alrighty, I think all comments have been addressed @realbigsean :pray:

**Comment by realbigsean** [comment]
> Presently we have (1) and I think that's ideal since we don't risk failing on block production. The downside of using (1) over (2) is that our head might bounce around a little from the EL perspective, which is safe but not ideal.

Yes that's true, I guess in block production it's just most important to build on top of `canonical_head` full stop.

File: beacon_node/beacon_chain/tests/payload_invalidation.rs
Comment (realbigsean) [inline beacon_node/beacon_chain/tests/payload_invalidation.rs]:
```suggestion
        prev_randao: *head
```

File: beacon_node/beacon_chain/src/beacon_chain.rs
Comment (realbigsean) [inline beacon_node/beacon_chain/src/beacon_chain.rs]:
```suggestion
            prev_randao: head.random,
```

**Review by realbigsean — COMMENTED** [review]
I tried to resolve a merge conflict via github, but there are a couple extra places that need re-naming.

**Review by realbigsean — APPROVED** [review]
Ok this looks good to me :slightly_smiling_face:

**Comment by paulhauner** [comment]
It turns out there's a bit of an edge-case with the transition block that isn't covered here. I've documented it in #3058.

I think we should merge this PR as-is since it's a significant improvement and has no issues before/after the transition block. We can come back for #3058.

**Comment by paulhauner** [comment]
Alrighty, I've been testing this on `merge-devnet-5` with our 1k validators. It's looking good there so I'm happy to merge.

There are two changes since last review:

1. Add in the changes from #3057
2. Add some metrics.

You can see these changes here: https://github.com/paulhauner/lighthouse/compare/76a1d1d...d029c7b

If I can get an approval for these changes then I'll merge :)

**Comment by paulhauner** [comment]
This is failing CI due to Geth changing some CLI flags. See https://github.com/sigp/lighthouse/pull/3046#issuecomment-1061292509

**Comment by paulhauner** [comment]
I've also added [4811406](https://github.com/sigp/lighthouse/pull/3043/commits/48114062a8f284573b6fe2a9980dd8f1d4048303) after observing a panic:

```
thread 'tokio-runtime-worker' panicked at 'Cannot start a runtime from within a runtime. This happens because a function (like `block_on`) attempted to block the current thread while the thread is being used to drive asynchronous tasks.'
```

That commit has approval from @AgeManning.

**Comment by paulhauner** [comment]
Sorry for the noise here. As you can see, I found some bugs whilst testing with our validators on `merge-devnet-5`.

I think this will require additional review, sorry @realbigsean. Unfortunately the diff between the previously-approved changes and the current commit includes a merge of `unstable` (I had to resolve a conflict, sorry).

We should be good to go after this round of review, this branch is performing well presently :rocket:

File: beacon_node/beacon_chain/tests/payload_invalidation.rs
Comment (realbigsean) [inline beacon_node/beacon_chain/tests/payload_invalidation.rs]: Is this intentionally left or a relic of debugging?
```suggestion
```

File: beacon_node/execution_layer/src/metrics.rs
Comment (realbigsean) [inline beacon_node/execution_layer/src/metrics.rs]:
```suggestion
        "Duration between an fcU call with PayloadAttributes and when the block should be produced",
```

or `a forkChoiceUpdated`

**Review by realbigsean — APPROVED** [review]
New changes look good to me, left a couple nits.  Resolved a merge conflict in `Cargo.lock` and a `Cargo.toml` as well to try to get CI running

**Comment by paulhauner** [comment]
Thanks for the review and CI fix @realbigsean :pray: 

I'll see how this goes on CI and the testnets and merge if all goes well.

**Comment by paulhauner** [comment]
Looks like windows tests are just being dodgy again :(

Everything is looking good on the testnet nodes :tada: 

bors r+

**Pull request closed without merging**
